src/otx/core/data/entity/detection.py

```diff
@@ -0,0 +1,72 @@
+"""Module for OTX detection data entities."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from otx.core.types.task import OTXTaskType
+
+from .base import OTXBatchDataEntity, OTXBatchPredEntity, OTXDataEntity, OTXPredEntity
+
+if TYPE_CHECKING:
+    from torch import LongTensor
+    from torchvision import tv_tensors
+
+
+@dataclass
+class DetDataEntity(OTXDataEntity):
+    """Data entity for detection task.
+
+    :param bboxes: Bbox annotations as top-left-bottom-right
+        (x1, y1, x2, y2) format with absolute coordinate values
+    :param labels: Bbox labels as integer indices
+    """
+
+    @property
+    def task(self) -> OTXTaskType:
+        """OTX Task type definition."""
+        return OTXTaskType.DETECTION
+
+    bboxes: tv_tensors.BoundingBoxes
+    labels: LongTensor
+
+
+@dataclass
+class DetPredEntity(DetDataEntity, OTXPredEntity):
+    """Data entity to represent the detection model output prediction."""
+
+
+@dataclass
+class DetBatchDataEntity(OTXBatchDataEntity[DetDataEntity]):
```

Comment:
`DataEntity` class seems to be designed for a specific use case. However, it's unclear if collating dictionaries could serve the same purpose. 
If we require four entities for each task, this could potentially impact scalability. Could we consider alternative approaches?

Reply:
Again, it's the design principal I proposed. There will be no more unstructured data objects in OTX v2.

I don't get your point for the scalability. These entities are corresponded to `OTXTaskType` 1-by-1. I don't think that `OTXTaskType` will scale to the enormous number in the future such as 10,000. Rather, unstructured data types make scaling new model development difficult. As you know, the number of models are always greater than the number of task type.

Reply:
While strict data structures can be useful, the necessity of DetPredEntity and DetBatchPredEntity, which are empty, is unclear. It's also not immediately apparent why a data structure like DetBatchDataEntity should contain a collate_fn function.

As a PyTorch user, I understand that `collate_fn` is used for data loaders, but its placement here is not intuitive.

Ideally, a data structure should be used for type checking, data storage, and serving as interfaces between different data structures (such as converting data types).

If these elements were more clearly defined, it would likely improve understanding. However, at present, the docstrings, naming conventions, and intentions seem ambiguous.

Reply:
Did you read the code or did you take your time to understand this design fully?

`DetPredEntity` and `DetBatchPredEntity` are not empty. They inherit `DetDataEntity`, `DetBatchDataEntity`, and `OTXPredEntity` at the same time. This means that they have one more degree of freedom to represent the model prediction score.

https://github.com/openvinotoolkit/training_extensions/blob/b8d8654ac4e846af0d0c4f1b0e971c5332ff6932/src/otx/core/data/entity/base.py#L81-L84

https://github.com/openvinotoolkit/training_extensions/blob/b8d8654ac4e846af0d0c4f1b0e971c5332ff6932/src/otx/core/data/entity/detection.py#L35-L37

You see here that `DetDataEntity` has capability to represent the model output prediction.

***
You are a code reviewer. However, it does not means that you have a super privilege to leave any comment what you want. I cannot understand why don't you take your time to understand first and get a review to move forward our product in a good way. Do you think that your time is only valuable one? How about my time? Don't you care about it? I have to answer again and again to you for the main concepts which I did such presentations, made slides, and shared with you guys already.

Did you think the following is the best thing you did as a reviewer?
![image](https://github.com/openvinotoolkit/training_extensions/assets/26541465/8f6abbb7-486e-4493-a4c0-6e6944a22ac3)

This directory is not really empty and it has a simple example for the future custom algo development, `hierarchical_cls_head.py` . Didn't you really catch it? I'm feeling that almost all of your comments are like this.

If I missed to deliver my design concepts fully to you, yes, sorry for that. If you want I'd like to explain it again to you until letting you understand fully. Please let me know anytime.
***

> And it seems there are better data structures abstract InstanceData and DetDataSample from MMEngine, and Instances from Detectron2. Why not reuse their data structure?

We are trying to go away from the hard coupling on MMEngine in OTX v1. That's why we do OTX v2. From now on, I will just resolve your comments about the main concepts of this design. If you have any other opinion or design concept, please prepare yours and discuss it with the OTX product owner, as same as I did.

Reply:
Hi, 
I haven't followed and understood all of your comments, but I think this conversation has crossed the line a bit from what I think should happen in PR.
While it is sometimes emotional for PR owners and reviewers to respond to each other logically, it should be minimized, and I think this conversation reflects poorly on those who will be commenting on this PR in the future.

Given that this is a place for all to see, I think the conversation should be edited or deleted. What do you think?